### PR TITLE
New rule: mixin and function named arguments

### PR DESCRIPTION
**Pull request by diego-codes**
Implements #181 

Two separate rules: `at-function-named-parameters` and `at-mixin-named-parameters` that implement the similar logic.

### CODE REVIEW

**Comment by kristerkari** [comment]
Thanks! I'll try to get this reviewed asap :)

File: src/rules/at-function-named-arguments/README.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
+# at-function-named-arguments
+
+Require named parameters in SCSS function call rule.
+
+```scss
+animation: animation($duration: 250ms) {
+//                 ↑
```
Comment (kristerkari) [inline src/rules/at-function-named-arguments/README.md]: Small thing, but this arrow should point to the `$`.

File: src/rules/at-function-named-arguments/README.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+```scss
+.foo {
+  border: reset($value: 20, 'bar', $color: #FFF);
+}
+```
+
+```scss
+.foo {
+  animation: animation($duration: 250ms);
```
Comment (kristerkari) [inline src/rules/at-function-named-arguments/README.md]: Why is this considered a warning when the rule is set to `always-multiple-arguments`?

File: src/rules/at-mixin-named-arguments/README.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+```scss
+.foo {
+  @include reset($value: 20, 'bar', $color: #FFF);
+}
+```
+
+```scss
+.foo {
+  @include animation($duration: 250ms);
```
Comment (kristerkari) [inline src/rules/at-mixin-named-arguments/README.md]: Same question as with Sass functions: should this be a warning if the rule is set to `always-multiple-arguments´?

**Comment by kristerkari** [comment]
Thanks @diego-codes! Sorry that it took me so long to review this, but I've been quite busy this week.

I added a couple of comments for things that I was unsure of. If you could check them out and after that I can merge this PR and release a new version.

Good job :+1:

File: src/rules/at-function-named-arguments/README.md (hunk 2 above)
Comment (diego-codes) [reply src/rules/at-function-named-arguments/README.md]: That’s is a good point. Would `only-multiple-arguments` make sense?

Comment (kristerkari) [reply src/rules/at-function-named-arguments/README.md]: I did a search in `stylelint` repo, but I didn't see `only-` prefix used in any of the rules, so maybe we should also try to keep things consistent.

I think that we can keep the name as it is right now, or then do something like `"always"` with a secondary option `{ except: ["single-argument"] }`, but I think that that might just be more confusing.

I'll need to sleep the night first to be able to think clearly :)

Comment (diego-codes) [reply src/rules/at-function-named-arguments/README.md]: Hey, sorry I'm just getting back to this. I agree it's important to keep things consistent. 

If we add the secondary `{ "except": ["single-argument"] }` option, would using it along with `"never"` mean that single arguments have to be named?

Looking at `stylelint`'s guidelines another approach can be `{ "ignoreSingleArgument": true }` which basically would avoid checking single arguments for either `"never"` or `"always"`. I'm not sure that I like this either, but it's more clear then `"except"` in my opinion.

What do you think?

Comment (kristerkari) [reply src/rules/at-function-named-arguments/README.md]: Using `"ignore"` could work, so how about `{ "ignore": ["single-argument"] }`? That style would also be extendable in the future.

Comment (diego-codes) [reply src/rules/at-function-named-arguments/README.md]: That sounds good. I’ll make this change.

**Comment by diego-codes** [comment]
Hey @kristerkari, I removed the `"always-multiple-arguments"` primary option and added the `"ignore: ["single-argument"]"` secondary option based on our discussion. The number of tests basically doubled which may be excessive...

Let me know what you think!

**Comment by kristerkari** [comment]
@diego-codes  Awesome! I would not be too worried about having a lot of tests. If you want to reduce duplication, you can use `.concat` like this rule does https://github.com/kristerkari/stylelint-scss/blob/master/src/rules/double-slash-comment-empty-line-before/__tests__/index.js#L135-L148

Let me know if you still want to make any changes. If no changes are needed, I can merge this 👍

File: src/rules/at-function-named-arguments/index.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+const isScssVarRegExp = /^\$\S*/;
+
+export default function(expectation, options) {
+  return function(root, result) {
+    const validOptions = utils.validateOptions(
+      result,
+      ruleName,
+      {
+        actual: expectation,
+        possible: ["always", "never", "always-multiple-arguments"]
```
Comment (kristerkari) [inline src/rules/at-function-named-arguments/index.js]: ` "always-multiple-arguments"` should be remove from here.

File: src/rules/at-mixin-named-arguments/index.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+const isScssVarRegExp = /^\$\S*/;
+
+export default function(expectation, options) {
+  return function(root, result) {
+    const validOptions = utils.validateOptions(
+      result,
+      ruleName,
+      {
+        actual: expectation,
+        possible: ["always", "never", "always-multiple-arguments"]
```
Comment (kristerkari) [inline src/rules/at-mixin-named-arguments/index.js]: ` "always-multiple-arguments"` should be remove from here too.

File: src/rules/at-function-named-arguments/index.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    "Unexpected a named parameter in single argument function call"
+});
+
+const cssFunctions = [
+  "attr",
+  "calc",
+  "linear-gradient",
+  "radial-gradient",
+  "repeating-linear-gradient",
+  "repeating-radial-gradient"
```
Comment (kristerkari) [inline src/rules/at-function-named-arguments/index.js]: Oh, btw. this function list seems incomplete, here's the full one:
https://github.com/EtaiG/css-functions-list

Comment (kristerkari) [reply src/rules/at-function-named-arguments/index.js]: and maybe you could also add a test that tests a CSS function that uses camelCase, for example `translateZ`.

Comment (diego-codes) [reply src/rules/at-function-named-arguments/index.js]: Good call!

**Comment by diego-codes** [comment]
@kristerkari, I made the changes you requested. I think I'm okay with the tests as they are. I just added some to tests for CSS functions.

File: src/rules/at-function-named-arguments/index.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+  isNativeCssFunction
+} from "../../utils";
+import valueParser from "postcss-value-parser";
+
+export const ruleName = namespace("at-function-named-arguments");
+
+export const messages = utils.ruleMessages(ruleName, {
+  expected: "Expected a named parameter to be used in function call",
+  rejected: "Unexpected a named parameter in function call",
+  rejectedSingle:
```
Comment (kristerkari) [inline src/rules/at-function-named-arguments/index.js]: I had another look at the code and I noticed that `rejectedSingle` does not seem to be used anywhere. It's also in the other rule.

File: src/rules/at-function-named-arguments/__tests__/index.js [hunk 8]
```diff
@@ -0,0 +1,852 @@
...
+  reject: [
+    {
+      code: `
+      .b {
+        border: reset($value: 40px);
+      }
+      `,
+      line: 3,
+      column: 9,
+      message: messages.unexpected,
```
Comment (kristerkari) [inline src/rules/at-function-named-arguments/__tests__/index.js]: Here and in the lines below the message should be `messages.rejected` as there is no `messages.unexpected` defined.

This also happens in the the tests for the other rule.

**Comment by kristerkari** [comment]
Great work @diego-codes! I found still a couple of really small things, if you could fix those, then I can merge this and release a new version tomorrow. :)

**Comment by diego-codes** [comment]
@kristerkari, sorry about that. That was really sloppy editing. Thanks for catching it. Let me know if you see anything else. Otherwise, I think it's good to go!

**Pull request merged**